From 59cb9bcae7f90cd6723ee6d475129288d4c36629 Mon Sep 17 00:00:00 2001
From: Robert Mustacchi <rm@joyent.com>
Date: Thu, 31 Jan 2019 05:09:40 +0000
Subject: [PATCH] OS-7563 mdb deserves a modulus operator Reviewed by: John
 Levon <john.levon@joyent.com> Reviewed by: Patrick Mooney
 <patrick.mooney@joyent.com> Approved by: Patrick Mooney
 <patrick.mooney@joyent.com>

---
 usr/src/cmd/mdb/common/mdb/mdb_grammar.y | 13 ++++++++++++-
 usr/src/cmd/mdb/common/mdb/mdb_lex.l     |  4 ++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/usr/src/cmd/mdb/common/mdb/mdb_grammar.y b/usr/src/cmd/mdb/common/mdb/mdb_grammar.y
index 97ea454627..05171acdef 100644
--- a/usr/src/cmd/mdb/common/mdb/mdb_grammar.y
+++ b/usr/src/cmd/mdb/common/mdb/mdb_grammar.y
@@ -26,7 +26,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.  All rights reserved.
+ * Copyright (c) 2019, Joyent, Inc.  All rights reserved.
  */
 
 #include <mdb/mdb_types.h>
@@ -112,6 +112,7 @@ yyexpand(int val)
 %left	MDB_TOK_LSHIFT MDB_TOK_RSHIFT
 %left	'-' '+'
 %left	'*' '%' '#'
+%left	MDB_TOK_MODULUS
 
 %right	MDB_COR_VALUE
 %right	MDB_OBJ_VALUE
@@ -309,6 +310,16 @@ expression:	expression '+' expression { $$ = $1 + $3; }
 			$$ = (intmax_t)$1 / (intmax_t)$3;
 		}
 
+	|	expression MDB_TOK_MODULUS expression {
+
+			if ($3 == 0UL)
+				yyerror("attempted to divide by zero");
+
+			$$ = $1 % $3;
+		}
+
+
+
 	|	expression '&' expression { $$ = $1 & $3; }
 	|	expression '|' expression { $$ = $1 | $3; }
 	|	expression '^' expression { $$ = $1 ^ $3; }
diff --git a/usr/src/cmd/mdb/common/mdb/mdb_lex.l b/usr/src/cmd/mdb/common/mdb/mdb_lex.l
index e777039c7b..7723f00f9e 100644
--- a/usr/src/cmd/mdb/common/mdb/mdb_lex.l
+++ b/usr/src/cmd/mdb/common/mdb/mdb_lex.l
@@ -28,6 +28,7 @@
 /*
  * Copyright 2011 Nexenta Systems, Inc. All rights reserved.
  * Copyright (c) 2017 by Delphix. All rights reserved.
+ * Copyright 2019, Joyent, Inc.
  */
 
 #include <sys/types.h>
@@ -131,6 +132,9 @@ RGX_COMMENT	"//".*\n
 <S_INITIAL>"!="	|
 <S_EXPR>"!="	return (MDB_TOK_NOTEQUAL); /* Inequality operator */
 
+<S_INITIAL>"%%"	|
+<S_EXPR>"%%"	return (MDB_TOK_MODULUS); /* Modulus operator */
+
 <S_INITIAL>"!"	|
 <S_FMTLIST>"!"	|
 <S_ARGLIST>"!"	{
-- 
2.21.0

